x-pack/plugins/apm/server/lib/helpers/setup_request.ts

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+/* tslint:disable no-console */
+import { SearchParams, SearchResponse } from 'elasticsearch';
+import moment from 'moment';
+import { Url } from 'url';
+import { StringMap } from '../../../typings/common';
+
+function decodeEsQuery(esQuery?: string) {
+  return esQuery ? JSON.parse(decodeURIComponent(esQuery)) : null;
+}
+
+// TODO: get these from hapi
+interface Request {
+  query: StringMap;
+  server: any;
+  method: string;
+  url: Url;
+}
+
+interface ServerConfig {
+  get: (key: string) => string;
+}
+
+type Client<T> = (type: string, params: SearchParams) => SearchResponse<T>;
+
+export interface Setup<T = any> {
+  start: number;
+  end: number;
+  esFilterQuery: string;
+  client: Client<T>;
+  config: ServerConfig;
+}
+
+export function setupRequest(req: Request, reply: (setup: Setup) => void) {
+  const cluster = req.server.plugins.elasticsearch.getCluster('data');
+
+  function client<T>(type: string, params: SearchParams): SearchResponse<T> {
```

Comment:
I would like to re-use the types from line 29:

https://github.com/elastic/kibana/blob/59734da6baadfe2c087e78865045c30ed2f17c79/x-pack/plugins/apm/server/lib/helpers/setup_request.ts#L29

Reply:
Does this work?
```ts
const client: Client<any> = (type, params) => {
  if (req.query._debug) {
    console.log(`DEBUG ES QUERY:`);
    console.log(
      `${req.method.toUpperCase()} ${req.url.pathname} ${JSON.stringify(
        req.query
      )}`
    );
    console.log(`GET ${params.index}/_search`);
    console.log(JSON.stringify(params.body, null, 4));
  }
  return cluster.callWithRequest(req, type, params);
};
```

Reply:
Or can use `function (type, params) {` instead of an arrow func if you want.

Reply:
Yes, neat! Only drawback is that it doesn't allow us to narrow down the generic argument, so the return value will always be `any`. I'll leave as-is for now, and maybe we'll find a better solution.